src/compiler/parser.ts

```diff
@@ -2717,6 +2745,31 @@ module ts {
             return expr;
         }
 
+        function parseBindingElementInitializer(inParameter: boolean): Expression {
+            // BindingElement[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingPattern[?Yield,?Await,GeneratorParameter] Initializer[In]opt
+            //      [+AsyncParameter] BindingPattern[?Yield,?Await,AsyncParameter] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingPattern[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            // SingleNameBinding[Yield,GeneratorParameter,Await,AsyncParameter] : 
+            //      [+GeneratorParameter] BindingIdentifier[Yield] Initializer[In]opt
+            //      [+AsyncParameter] BindingIdentifier[Await] Initializer[In]opt
+            //      [~GeneratorParameter,~AsyncParameter] BindingIdentifier[?Yield,?Await] Initializer[In,?Yield,?Await]opt
+            var explicitFlags = ParserContextFlags.None;
+            var ambientFlags = ParserContextFlags.None;
+            if (!inGeneratorParameterContext() && !inAsyncParameterContext()) {

```

Comment:
Even those GP and AP should be mutually-exclusive, I'm updating the various places where we make decisions based on those flags to something more like:

```TypeScript
                    // If [GeneratorParameter] is *set* we do not set [Yield] so it will be cleared by setContextParameters.
                    // If [GeneratorParameter] is *cleared* we pass the ambient value for [Yield].
                    if (!inGeneratorParameterContext()) {
                        ambientFlags |= ParserContextFlags.Yield;
                    }

                    // If [AsyncParameter] is *set* we do not set [Await] so it will be cleared by setContextParameters.
                    // If [AsyncParameter] is *cleared* we pass the ambient value for [Await].
                    if (!inAsyncParameterContext()) {
                        ambientFlags |= ParserContextFlags.Await;
                    }
```